cpp/src/io/statistics/column_stats.cu

```diff
@@ -0,0 +1,755 @@
+    if (!isnan(v)) vsum += v;
+    v = SHFL_XOR(vsum, 4);
+    if (!isnan(v)) vsum += v;
+    v = SHFL_XOR(vsum, 8);
+    if (!isnan(v)) vsum += v;
+    v = SHFL_XOR(vsum, 16);
+    if (!isnan(v)) vsum += v;
+    return vsum;
+}
+
+
+inline __device__ string_stats WarpReduceMinString(const char *smin, uint32_t lmin)
+{
+    uint32_t len = SHFL_XOR(lmin, 1);
+    const char *ptr = reinterpret_cast<const char *>(SHFL_XOR(reinterpret_cast<uintptr_t>(smin), 1));
+    if (!smin || (ptr && nvstr_is_lesser(ptr, len, smin, lmin))) {
+        smin = ptr;
+        lmin = len;
+    }
+    len = SHFL_XOR(lmin, 2);
+    ptr = reinterpret_cast<const char *>(SHFL_XOR(reinterpret_cast<uintptr_t>(smin), 2));
+    if (!smin || (ptr && nvstr_is_lesser(ptr, len, smin, lmin))) {
+        smin = ptr;
+        lmin = len;
+    }
+    len = SHFL_XOR(lmin, 4);
+    ptr = reinterpret_cast<const char *>(SHFL_XOR(reinterpret_cast<uintptr_t>(smin), 4));
+    if (!smin || (ptr && nvstr_is_lesser(ptr, len, smin, lmin))) {
+        smin = ptr;
+        lmin = len;
+    }
+    len = SHFL_XOR(lmin, 8);
+    ptr = reinterpret_cast<const char *>(SHFL_XOR(reinterpret_cast<uintptr_t>(smin), 8));
+    if (!smin || (ptr && nvstr_is_lesser(ptr, len, smin, lmin))) {
+        smin = ptr;
+        lmin = len;
+    }
+    len = SHFL_XOR(lmin, 16);
+    ptr = reinterpret_cast<const char *>(SHFL_XOR(reinterpret_cast<uintptr_t>(smin), 16));
+    if (!smin || (ptr && nvstr_is_lesser(ptr, len, smin, lmin))) {
+        smin = ptr;
+        lmin = len;
+    }
+    return {smin, lmin};
+}
+
+inline __device__ string_stats WarpReduceMaxString(const char *smax, uint32_t lmax)
+{
+    uint32_t len = SHFL_XOR(lmax, 1);
+    const char *ptr = reinterpret_cast<const char *>(SHFL_XOR(reinterpret_cast<uintptr_t>(smax), 1));
+    if (!smax || (ptr && nvstr_is_greater(ptr, len, smax, lmax))) {
+        smax = ptr;
+        lmax = len;
+    }
+    len = SHFL_XOR(lmax, 2);
+    ptr = reinterpret_cast<const char *>(SHFL_XOR(reinterpret_cast<uintptr_t>(smax), 2));
+    if (!smax || (ptr && nvstr_is_greater(ptr, len, smax, lmax))) {
+        smax = ptr;
+        lmax = len;
+    }
+    len = SHFL_XOR(lmax, 4);
+    ptr = reinterpret_cast<const char *>(SHFL_XOR(reinterpret_cast<uintptr_t>(smax), 4));
+    if (!smax || (ptr && nvstr_is_greater(ptr, len, smax, lmax))) {
+        smax = ptr;
+        lmax = len;
+    }
+    len = SHFL_XOR(lmax, 8);
+    ptr = reinterpret_cast<const char *>(SHFL_XOR(reinterpret_cast<uintptr_t>(smax), 8));
+    if (!smax || (ptr && nvstr_is_greater(ptr, len, smax, lmax))) {
+        smax = ptr;
+        lmax = len;
+    }
+    len = SHFL_XOR(lmax, 16);
+    ptr = reinterpret_cast<const char *>(SHFL_XOR(reinterpret_cast<uintptr_t>(smax), 16));
+    if (!smax || (ptr && nvstr_is_greater(ptr, len, smax, lmax))) {
+        smax = ptr;
+        lmax = len;
+    }
+    return { smax, lmax };
+}
+
+
+void __device__ gatherIntColumnStats(stats_state_s *s, statistics_dtype dtype, uint32_t t)
```

Comment:
libcudf has a lot of reduction/statistics machinery built already. I see that this code aims to combine multiple statistical computations in a single kernel here. However there is a lot of code duplication because every type has a dedicated function (largely the same) rather than using generic code.

Reply:
Also, the functionality of this code needs to be documented / explained.

Reply:
Added basic documentation. There are int, float and string paths, each doing something similar but different enough to warrant a separate implementation.